src/lib/operators/table_scan_implementation.hpp

```diff
@@ -0,0 +1,362 @@
+class TableScanImplementation : public TableScanBaseImplementation {
+ public:
+  TableScanImplementation(const std::shared_ptr<const Table> table, const ColumnID column_id, const ScanType scan_type,
+                          const AllTypeVariant search_value)
+      : TableScanBaseImplementation(table, column_id, scan_type, search_value),
+        _typed_search_value(type_cast<T>(_search_value)),
+        _result_table(std::make_shared<Table>()) {
+    ColumnID column_count = static_cast<ColumnID>(_table->column_count());
+    for (ColumnID column_id = ColumnID(0); column_id < column_count; ++column_id) {
+      _result_table->add_column(_table->column_name(column_id), _table->column_type(column_id));
+    }
+
+    _scan_type_comparator = _get_scan_type_comparator();
+  }
+
+  std::shared_ptr<const Table> on_execute() override {
+    ChunkID num_chunks = _table->chunk_count();
+    for (ChunkID chunk_id{0}; chunk_id < num_chunks; chunk_id++) {
+      const Chunk& chunk = _table->get_chunk(chunk_id);
+      _process_chunk(chunk);
+      _finish_current_chunk_offsets(chunk_id, chunk);
+    }
+
+    return _result_table;
+  }
+
+ protected:
+  // Exctract the relevant segment from the chunk and pass it to _process_segment
+  void _process_chunk(const Chunk& source_chunk) {
+    const auto source_segment = source_chunk.get_segment(_column_id);
+
+    const auto dictionary_segment = std::dynamic_pointer_cast<DictionarySegment<T>>(source_segment);
+    if (dictionary_segment) {
+      _process_segment(dictionary_segment);
+      return;
+    }
+
+    const auto reference_segment = std::dynamic_pointer_cast<ReferenceSegment>(source_segment);
+    if (reference_segment) {
+      _process_segment(reference_segment);
+      return;
+    }
+
+    const auto value_segment = std::dynamic_pointer_cast<ValueSegment<T>>(source_segment);
+    if (value_segment) {
+      _process_segment(value_segment);
+      return;
+    }
+
+    throw std::logic_error("Unhandled segment type in _process_chunk");
+  }
+
+  // -----------------
+  // These methods go through the segment and add all relevant values to _chunk_offsets_to_add_to_result_table
+  // OR set _add_all_chunk_offsets to prevent copying the whole list
+  void _process_segment(std::shared_ptr<ValueSegment<T>> segment) {
+    const auto& values = segment->values();
+    const size_t values_size = values.size();
+
+    for (size_t value_index = 0; value_index < values_size; ++value_index) {
+      if (_scan_type_comparator(values[value_index])) {
+        _chunk_offsets_to_add_to_result_table.push_back(value_index);
+      }
+    }
+  }
+
+  void _process_segment(std::shared_ptr<DictionarySegment<T>> segment) {
+    const DictionarySegmentProcessingFlags flags(segment, _scan_type, _typed_search_value);
+
+    if (flags.add_none) {
+      return;
+    }
+
+    if (flags.add_all) {
+      _add_all_chunk_offsets = true;
+    } else {
+      const auto& attribute_vector = *(segment->attribute_vector());
```

Comment:
You could have just worked on a pointer here. But the only benefit would be that you could do it without dereferencing a pointer here.